Code/GraphMol/MarvinParse/MarvinParser.cpp

```diff
@@ -0,0 +1,2336 @@
+
+#include <string>
+#include <exception>
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <sstream>
+#include <iomanip>
+#include <cfloat>
+
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+#include <boost/foreach.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/format.hpp>
+#include <RDGeneral/BoostEndInclude.h>
+
+#include <GraphMol/FileParsers/FileParsers.h>
+#include <GraphMol/FileParsers/MolSGroupParsing.h>
+#include <GraphMol/FileParsers/MolFileStereochem.h>
+#include "MarvinParser.h"
+#include "MarvinDefs.h"
+#include <GraphMol/Conformer.h>
+
+#include <GraphMol/RDKitQueries.h>
+#include <GraphMol/StereoGroup.h>
+#include <GraphMol/SubstanceGroup.h>
+
+#include "MarvinParser.h"
+
+#include <RDGeneral/StreamOps.h>
+#include <RDGeneral/FileParseException.h>
+#include <RDGeneral/BadFileException.h>
+#include <RDGeneral/LocaleSwitcher.h>
+
+#ifdef RDKIT_USE_BOOST_REGEX
+#include <boost/regex.hpp>
+using boost::regex;
+using boost::regex_match;
+using boost::smatch;
+// using boost::algorithm>;
+
+#else
+#include <regex>
+using std::regex;
+using std::regex_match;
+using std::smatch;
+#endif
+
+using namespace RDKit::SGroupParsing;
+
+namespace RDKit {
+
+/*
+      Imports the Marvin-specific dialect of CML (Chemical Markup Language) and
+   converts it to datastructures that are compatible with Molfile, RXNfile, and
+   Molfile complemented with canvas objects.
+*/
+class MarvinCMLReader {
+ public:
+  MarvinCMLReader(){};
+
+  ~MarvinCMLReader(){};
+
+  RWMol *parseMolecule(boost::property_tree::ptree molTree,
+                       bool sanitize = false, bool removeHs = false) {
+    MarvinMol *marvinMol = nullptr;
+    RWMol *mol = nullptr;
+    boost::property_tree::ptree molSection;
+
+    boost::property_tree::ptree molOrRxn;
+
+    try {
+      molSection = molTree.get_child("cml.MDocument.MChemicalStruct.molecule");
+    } catch (const std::exception &e) {
+      try {
+        molSection = molTree.get_child("cml.MDocument");
+        return new RWMol();
+      } catch (const std::exception &e) {
+        try {
+          molSection = molTree.get_child("cml");
+          return new RWMol();
+        } catch (const std::exception &e) {
+          throw FileParseException("Expected \"molecule\" in MRV file");
+        }
+      }
+    }
+
+    try {
+      marvinMol = (MarvinMol *)parseMarvinMolecule(molSection);
+      // marvinMol->convertFromSuperAtoms();
+      // marvinMol->processMulticenterSgroups();
+      // marvinMol->expandMultipleSgroups();
+      marvinMol->prepSgroupsForRDKit();
+
+      mol = parseMolecule(marvinMol, sanitize, removeHs);
+
+      delete marvinMol;
+
+      return mol;
+    } catch (const std::exception &e) {
+      delete mol;
+
+      delete marvinMol;
+
+      throw;
+    }
+  }
+
+  ChemicalReaction *parseReaction(boost::property_tree::ptree rxnTree,
```

Comment:
The github interface won't let me make this suggestion directly, but I'd suggest replacing the contents of this function with the following simplification:
```
    std::unique_ptr<ChemicalReaction> rxn{new ChemicalReaction()};
    rxnTree = rxnTree.get_child("cml.MDocument.MChemicalStruct.reaction");
    std::unique_ptr<MarvinReaction> marvinReaction{
        parseMarvinReaction(rxnTree, documentTree)};
    marvinReaction->prepSgroupsForRDKit();

    // get each reactant
    for (auto mol : marvinReaction->reactants) {
      rxn->addReactantTemplate(
          ROMOL_SPTR(parseMolecule(mol, sanitize, removeHs)));
    }

    // get each agent
    for (auto mol : marvinReaction->agents) {
      rxn->addAgentTemplate(ROMOL_SPTR(parseMolecule(mol, sanitize, removeHs)));
    }

    // get each product
    for (auto mol : marvinReaction->products) {
      rxn->addProductTemplate(
          ROMOL_SPTR(parseMolecule(mol, sanitize, removeHs)));
    }

    // convert atoms to queries:
    for (auto mol : rxn->getReactants()) {
      for (auto atom : mol->atoms()) {
        QueryOps::replaceAtomWithQueryAtom(static_cast<RWMol *>(mol.get()),
                                           atom);
      }
    }
    for (auto mol : rxn->getProducts()) {
      for (auto atom : mol->atoms()) {
        QueryOps::replaceAtomWithQueryAtom(static_cast<RWMol *>(mol.get()),
                                           atom);
      }
    }

    return rxn.release();
```